File: maximalSquare3.java
//Time complexity: O(mn)
//Space complexity: O(m)
//DP approach using adiitional array instead of matrix

class maximalSqr {
    public int maximalSquare(char[][] matrix) {
        if(matrix == null || matrix.length == 0)
            return 0;
        
        int m = matrix.length, n = matrix[0].length;
        int max = 0, prev = 0;
        int[] dp = new int[n + 1];
        
        for(int i = 1; i < m + 1; i++){
            for(int j = 1; j < n + 1; j++){
                if(matrix[i - 1][j - 1] == '1'){
                    int temp = dp[j];
                    dp[j] = Math.min(dp[j], Math.min(dp[j - 1], prev)) + 1;
                    prev = temp;
                    
                    
                }
                else{
                    dp[j] = 0;
                }
                max = Math.max(max, dp[j]);
            }
        }
        
        return max*max;
    }
}

File: maximal_square3.py
# // Time Complexity :O((m*n)
# // Space Complexity :O(m)
# // Did this code successfully run on Leetcode :yes
# // Any problem you faced while coding this :no


# // Your code here along with comments explaining your approach



class Solution:
    
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        m=len(matrix)
        n=len(matrix[0])
        ma=0
        dp=[0 for i in range(n+1)]
        temp=0
        temp2=0
        print(dp)
        for i in range(1,m+1):
            
            for j in range(1,n+1):
                temp=dp[j]
                if matrix[i-1][j-1]=='1':
                    print(matrix[i-1][j-1])
                    print(dp[j])
                    print(temp)
                    dp[j]=min(dp[j],dp[j-1],temp2)+1
                else:
                    dp[j]=0
                temp2=temp
                ma=max(ma,dp[j])
            print(dp)
        return ma*ma
        
                    

        
                    
                        
                        
        